lib/ui/compositing.dart

```diff
@@ -47,18 +46,16 @@ class Scene extends NativeFieldWrapperClass1 {
     if (width <= 0 || height <= 0) {
       throw Exception('Invalid image dimensions.');
     }
-    return _futurize((_Callback<Image?> callback) => _toImage(width, height, (_Image? image) {
-        if (image == null) {
-          callback(null);
-        } else {
-          callback(Image._(image, image.width, image.height));
-        }
-      }),
-    );
+    final _Image image = _Image._();
+    final String? result =  _toImage(width, height, image);
+    if (result != null) {
+      throw PictureRasterizationException._(result);
+    }
+    return Future<Image>.value(Image._(image, image.width, image.height));
   }
 
   @FfiNative<Handle Function(Pointer<Void>, Uint32, Uint32, Handle)>('Scene::toImage')
-  external String? _toImage(int width, int height, _Callback<_Image?> callback);
+  external String? _toImage(int width, int height, _Image outImage);
```

Comment:
I'm not sure this is the right choice here.

The current API returns a host copy of the image every time, which can't get lost of GPU context gets lost and is safe to use in the background.

The new API fails in this scenario:

```dart
final image = await scene.toImage(40, 40);
// go to background
final data = await image.toByteData(...); // won't work.
```

I think we should try to merge these in the C++ layer for now.

Reply:
So we could let the APIs diverge then, but I'm not sure of the exact mechanics here. Since the texture registry is only available on the raster thread, don't we need to both flatten and rasterize on that thread in order to retain the contents?

We can always punt on this for now, but I'd like to figure out what the path forward is.

Reply:
i.e. can I tweak the toImage path to flatten/rasterize on the raster thread and still be safe here?

Reply:
`toImage` already posts a task to the raster thread. We'd just need to move the flattening of the layer tree to that task.

I think the path forward would be to refactor `Picture::RasterizeToImage` so that its guts live on `Scene::RasterizeToImage` and the scene version knows how to convert a layer tree to a display list on the raster thread (but also can still take a display list if the call is coming from Picture, which already has one).

Reply:
Got it